### ISSUE TITLE
Permissions for non-managed resources: Unexpected ability to query timestream

### ISSUE THREAD

**Issue by dmeehan1968** [sst/ion#213]
I'm adding SST/Ion to an existing CDK based application, and want to be able to access resources from the existing deployment from a NextJs application. The intention is to implement a dashboard UI that provides access to the existing resource, such as metrics stored in Timestream, service configurations in Dynamo etc. If all goes well I will import the resources into the SST application, but for now I just want to experiment with the frontend deployment.

My first test was to be able to query a Timestream table and produce a table of results in the UI. Timestream isn't a supported component type in SST, and anyway, the database/table already exists so I can't 'import' it as an SST managed resource, and all I need to be able to do is access it by name/ARN.

I was expecting to need to add appropriate permissions to the NextJs instance in the sst.config.ts, but I have been able to issue a query without doing so. I cannot see from the created IAM role/policies (`app-user-DashboardDefaultRole-region`) how the access is being granted, as this appears to be just the minimum required for SST access its resources (S3, Queue, Dynamo, lambda).

I have a single set of AWS credentials in \~/.aws/credentials which are for my AWS console user, and provides the necessary access for SST to deploy.

I'm a complete novice with SST (both CDK and Ion variants) so may not be approaching this correctly. In my CDK app that creates those resources, I use a least permissive approach as per AWS guidelines, so any existing resource such as a user or function is explicitly granted the permissions required (e.g. a function only gets Timestream Query permissions if that's all it needs). I don't think therefore that there can be an existing policy that would unintentionally grant access, other than the user that SST is using for deployment. The DashboardDefaultFunction has the DashboardDefaultRole assigned and the console permissions viewer only shows the expected resources and doesn't include Timestream.

This is my sst.config.ts file:

```ts
///
 
export default $config({
  app(input) {
    return {
      name: "replicated",
      removal: input?.stage === "production" ? "retain" : "remove",
      home: "aws",
    };
  },
  async run() {
    new sst.aws.Nextjs('Dashboard', {
      path: 'packages/dashboard',
    })
  },
});
```

**Comment by dmeehan1968** [comment]
One thing I realised is that when working locally, I was using `sst dev` instead of `sst dev next dev` and was running next in dev mode separately.  What I reported above was experienced both in `sst dev` mode, and after `sst deploy` when accessing the CF endpoint.

When I switch to `sst dev next dev`, I now get 

  > "Error: The operation to discover endpoint failed. Please retry, or provide a custom endpoint and disable endpoint discovery to proceed." 

which is what I had expected to encounter in the first place (without permissions).  

However, if I add the permissions for `timestream:DescribeEndpoints` and `timestream:Select` to my config, it still fails with the same message, and inspecting the DashboardDefaultRole does not show the timestream permissions.  

Here is the update Nextjs instance:

```ts
    new sst.aws.Nextjs('Dashboard', {
      path: 'packages/dashboard',
      permissions: [
        { actions: ['timestream:DescribeEndpoints'], resources: ['*'] },
        { actions: ['timestream:Select'], resources: ['arn:aws:timestream:eu-west-1:<REDACTED>:database/<REDACTED>/table/<REDACTED>'] },
      ]
    })
```

I've tried a fresh deploy, but the IAM role doesn't seem to get updated with the permissions.

**Comment by dmeehan1968** [comment]
There is a slightly cleaner way of intercepting the server function to add the permissions:

```ts
    new sst.aws.Nextjs('Dashboard', {
      path: 'packages/dashboard',
      transform: {
        server: (fn) => {
          if (Array.isArray(fn.permissions)) {
            fn.permissions = [
              ...fn.permissions,
              { actions: ['timestream:DescribeEndpoints'], resources: ['*'] },
              { actions: ['timestream:Select'], resources: ['arn:aws:timestream:eu-west-1:<REDACTED>:database/<REDACTED>/table/<REDACTED>']
              },
            ]
          }
        }
      }
```

It seems that the transform function was designed without taking account of Pulumi handling values that are `T | Promise<T> | OutputInstance<T>`, which makes alterations tricky.  Pulumi contains a function `output` that is designed to unwrap any value and return an OutputInstance<T>, whilst allowing changes to be made using the `.apply()` method.

The transform function only allows `T | undefined`, which doesn't allow for unwrapping.  It would be better if the transform also allowed an `OutputInstance` so that you could then do:

```ts
{
  transform: {
    server: (args: FunctionArgs) => {
      return output(args).apply(args => {
          fn.permissions = [
            ...fn.permissions,
            { actions: ['timestream:DescribeEndpoints'], resources: ['*'] },
            { actions: ['timestream:Select'], resources: ['arn:aws:timestream:eu-west-1:<REDACTED>:database/<REDACTED>/table/<REDACTED>']
            },
          ]
      }
   }
}
```

**Comment by dmeehan1968** [comment]
A useful helper function to disambiguate whether the function property is an Output or a Promise to apply the changes more succinctly:

```ts
function apply<T>(input: Input<T> | undefined, fn: (input: T) => void) {
  if (Output.isInstance(input)) {
    input.apply(fn)
  } else if (input instanceof Promise) {
    input.then(input => {
      fn(input)
      return input
    })
  } else if (input !== undefined) {
    fn(input)
  }
}
```

Use like:

```ts
      transform: {
        server: (fn) => {
          const databaseName = '<REDACTED>'
          const tableName = '<REDACTED>'
          apply(fn.permissions, perms => {
            perms.push({ actions: ['timestream:DescribeEndpoints'], resources: ['*'] })
            perms.push({ actions: ['timestream:Select'], resources: [`arn:aws:timestream:eu-west-1:<REDACTED>:database/${databaseName}/table/${tableName}`] })
          })
          apply(fn.environment, env => {
            env.METRICS_DATABASE_NAME = databaseName
            env.METRICS_TABLE_NAME = tableName
          })
        }
      }
```

**Comment by JuniYadi** [comment]
> There is a slightly cleaner way of intercepting the server function to add the permissions:
> 
> ```ts
>     new sst.aws.Nextjs('Dashboard', {
>       path: 'packages/dashboard',
>       transform: {
>         server: (fn) => {
>           if (Array.isArray(fn.permissions)) {
>             fn.permissions = [
>               ...fn.permissions,
>               { actions: ['timestream:DescribeEndpoints'], resources: ['*'] },
>               { actions: ['timestream:Select'], resources: ['arn:aws:timestream:eu-west-1:<REDACTED>:database/<REDACTED>/table/<REDACTED>']
>               },
>             ]
>           }
>         }
>       }
> ```
> 
> It seems that the transform function was designed without taking account of Pulumi handling values that are `T | Promise<T> | OutputInstance<T>`, which makes alterations tricky. Pulumi contains a function `output` that is designed to unwrap any value and return an OutputInstance, whilst allowing changes to be made using the `.apply()` method.
> 
> The transform function only allows `T | undefined`, which doesn't allow for unwrapping. It would be better if the transform also allowed an `OutputInstance` so that you could then do:
> 
> ```ts
> {
>   transform: {
>     server: (args: FunctionArgs) => {
>       return output(args).apply(args => {
>           fn.permissions = [
>             ...fn.permissions,
>             { actions: ['timestream:DescribeEndpoints'], resources: ['*'] },
>             { actions: ['timestream:Select'], resources: ['arn:aws:timestream:eu-west-1:<REDACTED>:database/<REDACTED>/table/<REDACTED>']
>             },
>           ]
>       }
>    }
> }
> ```

we should put this on documentation

**Comment by fwang** [comment]
Should be fixed in [v0.0.324](https://github.com/sst/ion/releases/tag/v0.0.324).

Feel free to reopen if the issue persists.
